### ISSUE TITLE
Memory usage for reading of 268 MiB

### ISSUE THREAD

**Issue by VelizarVESSELINOV** [kinverarity1/lasio#101]
My feeling is that lasio is using too much memory. Please find a test program that I wrote to monitor the memory usage: https://github.com/VelizarVESSELINOV/DockerTest

I created a dummy LAS file with 0.5M rows and 51 mnemonics the result is 267.98 MiB file which is reasonable size (without being very large) for LAS file.

My expected maximum memory usage is numpy.ndarray which is most probably 1/8 of the LAS ASCII file.

I didn't expect the raw ASCII file to be put in memory. The result of `pympler` after `lasio.read`:

```
Reading file
                                types |   # objects |   total size
===================================== | =========== | ============
                          <class 'str |        4356 |    268.28 MB
                         <class 'list |        4214 |    395.56 KB
            <class 'lasio.las.LASFile |           1 |     35.26 KB
                          <class 'int |         778 |     21.27 KB
                    <class 'weakproxy |         176 |     13.75 KB
            <class 'collections._Link |         176 |     12.38 KB
                         <class 'dict |           9 |      8.25 KB
  <class 'lasio.las.OrderedDictionary |           3 |      6.43 KB
                <class 'numpy.ndarray |          51 |      3.98 KB
              <class 'lasio.las.Curve |          51 |      3.98 KB
         <class 'lasio.las.HeaderItem |          19 |      1.34 KB
                         <class 'type |           0 |    576     B
                        <class 'float |           5 |    120     B
                        <class 'tuple |           1 |     64     B
      function (_count_diff_hashable) |           0 |      0     B
```

The Mac Activity monitor is giving me memory usage of 3 GiB with 5.3 GiB virtual memory.

<img width="559" alt="screen shot 2015-08-23 at 7 59 46 pm" src="https://cloud.githubusercontent.com/assets/8209940/9432582/12e6830a-49da-11e5-8751-bca5ec8739d4.png">

The Python `memory_profiler` is giving line-by-line memory consumption of 500 MiB per read.
I have the feeling that they are not capturing the numpy allocations that are done in C or Fortan.

```
Line #    Mem usage    Increment   Line Contents
================================================
    31     27.8 MiB      0.0 MiB   @profile
    32                             def file_test(rows=500000, cols=50):
    33                                 "File test"
    34     27.8 MiB      0.0 MiB       print("Creating file with {} rows and {} columns".format(rows, cols))
    35     88.4 MiB     60.6 MiB       file = create_file(rows, cols)
    36     88.4 MiB      0.0 MiB       print("Size of the file: {:.2f} MiB".format(getsize(file) / (1024 * 1024)))
    37     88.4 MiB      0.0 MiB       print("Reading file")
    38     64.0 MiB    -24.4 MiB       sum1 = summarize(get_objects())
    39    551.9 MiB    487.9 MiB       las = read(file)
    40    524.4 MiB    -27.5 MiB       sum2 = summarize(get_objects())
    41    524.4 MiB      0.0 MiB       diff = get_diff(sum1, sum2)
    42    524.4 MiB      0.0 MiB       print_(diff)
    43                             
    44    529.3 MiB      4.9 MiB       for curve in las.curves:
    45    529.3 MiB      0.0 MiB           print("Name: {}, Min: {:.2f}, Mean: {:.2f}, Max: {:.2f}"
    46    529.3 MiB      0.0 MiB                 .format(curve.mnemonic, nanmin(curve.data), nanmean(curve.data),
    47    529.3 MiB      0.0 MiB                         nanmax(curve.data)))
    48                             
    49    257.3 MiB   -272.0 MiB       del las
    50    752.4 MiB    495.1 MiB       las = read(file)
    51    285.9 MiB   -466.5 MiB       del las
    52    757.6 MiB    471.7 MiB       las = read(file)
    53    290.0 MiB   -467.5 MiB       del las
    54    756.4 MiB    466.4 MiB       las = read(file)
    55    287.6 MiB   -468.8 MiB       del las
    56    287.6 MiB      0.0 MiB       print("Happy end")
```

The `psutil` process activity profiler is giving the plot below for 5 seconds interval:
![psutilprofile](https://cloud.githubusercontent.com/assets/8209940/9432651/e8a8f888-49da-11e5-9af3-d241469b7e43.png)
- cpu: in yellow almost 100% all the time (normal)
- rss: in magenta for Resident Set Size (equivalent to Mem Usage on Windows)
- vms: in dark blue for Virtual Memory Size

The plot is generated by https://github.com/VelizarVESSELINOV/DockerTest/blob/master/process_log.py

More than 2 GiB of memory usage for reading less than 0.3 GiB file it the opposite ratio that I expect.

I really like the library but for large file I need to use very expensive large memory nodes/laptop or you will be able to optimize the reading memory usage.

Please try to optimize the the reading memory usage as I mentioned in issue #50 with `numpy.fromiter` and please check with the code to reduce the memory usage. I done in the past several large CSV file readers in Python and managed to read several GiB of files without problems.

**Comment by kinverarity1** [comment]
Yep, no surprises as I haven't yet worked on this.

#50 is still open and a priority to fix as the reader code was written a while ago as the simplest option.

Thanks for the very detailed report.

**Comment by kinverarity1** [comment]
Addressed by #170 hopefully
